### PR TITLE
Fix bug. get thresholds broken for multi_label

### PR DESCRIPTION
**Pull request by cgnorthcutt**
Addresses the error found by @suneeta-mall in https://github.com/cleanlab/cleanlab/issues/263. Thanks Suneeta!

The bug was introduced  when the`get_confident_thresholds` method  was added to count.py (multi_label support was dropped the computation used does not work for multi_label -- this should be fixed by this PR).

In the future, more testing of multi_label labels should be added to our tests to catch errors like [this](https://github.com/cleanlab/cleanlab/issues/263).

### CODE REVIEW

File: cleanlab/count.py [hunk 1]
```diff
@@ -1108,12 +1111,31 @@ def get_confident_thresholds(labels: np.array, pred_probs: np.array) -> np.array
...
     confident_thresholds : np.array
       An array of shape ``(K,)``.
     """
-    confident_thresholds = np.array(
-        [np.mean(pred_probs[:, k][labels == k]) for k in range(pred_probs.shape[1])]
-    )
+
+    if multi_label:
+        # Compute unique number of classes K by flattening labels (list of lists)
+        K = len(np.unique([i for lst in labels for i in lst]))
```
Comment (jwmueller) [inline cleanlab/count.py:1132]: why not define `K = pred_probs.shape[1]` like is being done below when multi_label=false?

Comment (jwmueller) [reply cleanlab/count.py:1132]: otherwise I think anywhere num-classes are being computed based on num-unique-values, the docstring should explicitly state that all classes must be present in `labels` to avoid possible user confusion like we have seen in some Issues.

File: cleanlab/count.py [hunk 2]
```diff
@@ -1108,12 +1111,31 @@ def get_confident_thresholds(labels: np.array, pred_probs: np.array) -> np.array
...
+        # Compute thresholds = p(label=k | k in set of given labels)
+        k_in_l = np.array([[k in lst for lst in labels] for k in range(K)])
+        # The avg probability of class given that the label is represented.
+        confident_thresholds = [np.mean(pred_probs[:, k][k_in_l[k]]) for k in range(K)]
+    else:
+        confident_thresholds = np.array(
+            [np.mean(pred_probs[:, k][labels == k]) for k in range(pred_probs.shape[1])]
+        )
     return confident_thresholds
```
Comment (jwmueller) [inline cleanlab/count.py:1141]: seems like `confident_thresholds` is not always a np.array. I'd remove the `np.array(...)` wrapper from L1138 and instead put it at the return in L1141:

```
return np.array(confident_thresholds)
```

File: cleanlab/count.py [hunk 3]
```diff
@@ -425,7 +424,7 @@ class 0, 1, ..., K-1. `pred_probs` should have been computed using 3 (or
     # Estimate the probability thresholds for confident counting
     if thresholds is None:
         # P(we predict the given noisy label is k | given noisy label is k)
-        thresholds = [np.mean(pred_probs[:, k][labels == k]) for k in range(num_classes)]
+        thresholds = get_confident_thresholds(labels, pred_probs, multi_label=multi_label)
     thresholds = np.asarray(thresholds)
```
Comment (jwmueller) [inline cleanlab/count.py:428]: `thresholds = np.asarray(thresholds)` seems unnecessary if the `get_confident_thresholds()` properly returns a np.array like it says it does in its docstring.

File: cleanlab/count.py [hunk 4]
```diff
@@ -263,16 +263,15 @@ def _compute_confident_joint_multi_label(
 
     # Compute unique number of classes K by flattening labels (list of lists)
     K = len(np.unique([i for lst in labels for i in lst]))
```
Comment (jwmueller) [inline cleanlab/count.py:265]: If we're computing num_classes based on unique values, the docstring should explicitly state that all classes must be present in labels to avoid possible user confusion like we have seen in some Issues.

File: cleanlab/count.py (hunk 1 above)
Comment (jwmueller) [reply cleanlab/count.py:1132]: overall I think it's important to be consistent more than anything else. So num_classes should be either be defined based on `pred_probs.shape[1]` or num-unique-values throughout the whole package. 

The former is more computationally efficient, but has the disadvantage that existing code would need to be changed (eg. in thresholds computation) to account for potentially missing classes not present in `labels`.

File: cleanlab/internal/label_quality_utils.py [hunk 5]
```diff
@@ -35,19 +39,33 @@ def _subtract_confident_thresholds(labels: np.array, pred_probs: np.array) -> np
     pred_probs : np.array (shape (N, K))
       Predicted-probabilities in the same format expected by the `cleanlab.count.get_confident_thresholds()` method.
 
+    multi_label : bool, optional
+      multi_label in the same format expected by the `cleanlab.count.get_confident_thresholds()` method.
+
+    multi_label : bool, optional
+      If ``True``, labels should be an iterable (e.g. list) of iterables, containing a
+      list of labels for each example, instead of just a single label.
```
Comment (jwmueller) [inline cleanlab/internal/label_quality_utils.py:47]: multi_label arg is documented twice

**Review by jwmueller — CHANGES_REQUESTED** [review]
just have some minor changes/questions.

File: cleanlab/count.py [hunk 6]
```diff
@@ -263,16 +263,15 @@ def _compute_confident_joint_multi_label(
...
     if thresholds is None:
         # the avg probability of class given that the label is represented.
-        thresholds = [np.mean(pred_probs[:, k][k_in_l[k]]) for k in range(K)]
+        thresholds = get_confident_thresholds(labels, pred_probs, multi_label=True)
     # Create mask for every example if for each class, prob >= threshold
     pred_probs_bool = pred_probs >= thresholds
 
     # Compute confident joint
     # (no need to avoid collisions for multi-label, double counting is okay!)
+    k_in_l = np.array([[k in lst for lst in labels] for k in range(K)])
```
Comment (jwmueller) [inline cleanlab/count.py:274]: A nitpick is this code may be 2x as slow as the previous version for dataset with tons of labels, because you are computing this `k_in_l` object twice now (whereas previously it was only computed once), the first time being inside `get_confident_thresholds`

File: cleanlab/count.py (hunk 1 above)
Comment (jwmueller) [reply cleanlab/count.py:1132]: Perf optimization if we use num-unique-values to set num-classes:  then certain helper functions could take in `num_classes` as an optional argument to avoid having to necessarily recompute: `unique(labels)` which may be costly for massive dataset.

File: cleanlab/count.py (hunk 4 above)
Comment (suneeta-mall) [reply cleanlab/count.py:265]: I second this comment ^^. This is so true in multi-label and you are likely to run into this case give you explicitly don't ask for one hot encoded label for multi-label settings. 
I more I think about it, the more I feel, that it will be better to have a different abstraction for multi-label than trying to wedge it in via flag. At the same time, I appreciate that this support is experimental so all good! 

Left some thoughts here https://github.com/cleanlab/cleanlab/issues/263#issuecomment-1133885728 but I won't say my thoughts are well thought through just yet.

File: cleanlab/count.py (hunk 1 above)
Comment (cgnorthcutt) [reply cleanlab/count.py:1132]: done

File: cleanlab/count.py (hunk 3 above)
Comment (cgnorthcutt) [reply cleanlab/count.py:428]: Leave as is. User could have passed in thresholds... as a list. Note this occurs after the if statement.

File: cleanlab/count.py (hunk 4 above)
Comment (cgnorthcutt) [reply cleanlab/count.py:265]: clarified this in labels param description. Also the following tutorials PR helped address this: https://github.com/cleanlab/cleanlab/pull/279

File: cleanlab/internal/label_quality_utils.py (hunk 5 above)
Comment (cgnorthcutt) [reply cleanlab/internal/label_quality_utils.py:47]: fixed. thx @jwmueller

File: cleanlab/count.py (hunk 6 above)
Comment (cgnorthcutt) [reply cleanlab/count.py:274]: I'm going to allow this redundancy for now. The two options to fix are:
1. compute `k_in_l` beforehand and pass it into `get_confident_thresholds`, but its a weird thing to pass in and the code is much easier to read if its computed inside of `get_confident_thresholds`
2. return `k_in_l` also when calling `get_confident_thresholds` with `multi_label=true` but that's pretty weird behavior. I can optionally include it in the return tuple with another parameter, but this seems like a somewhat silly amount of extra code to save very little compute.

the computation here should be relatively negligible so going to leave as is for the purpose of easier-to-read simpler code.

Comment (cgnorthcutt) [reply cleanlab/count.py:274]: Easy for me to implement either solution if you think its preferable. I vote to leave as is for code simplicity, but can go either way. Agree its 2x speed, but if we get to the place where this is the bottleneck than I'll be extremely happy :)

File: cleanlab/count.py [hunk 7]
```diff
@@ -1108,12 +1107,31 @@ def get_confident_thresholds(labels: np.array, pred_probs: np.array) -> np.array
...
     -------
     confident_thresholds : np.array
       An array of shape ``(K,)``.
     """
-    confident_thresholds = np.array(
-        [np.mean(pred_probs[:, k][labels == k]) for k in range(pred_probs.shape[1])]
-    )
+
+    if multi_label:
+        # [0, 1, 2, ... num_classes - 1] Assumes all classses are represented in labels
```
Comment (jwmueller) [inline cleanlab/count.py:1127]: move this sentence to the docstring for `labels` arg?

File: cleanlab/count.py [hunk 8]
```diff
@@ -1108,12 +1107,31 @@ def get_confident_thresholds(labels: np.array, pred_probs: np.array) -> np.array
...
+        unique_classes = range(pred_probs.shape[1])
+        # Compute thresholds = p(label=k | k in set of given labels)
+        k_in_l = np.array([[k in lst for lst in labels] for k in unique_classes])
+        # The avg probability of class given that the label is represented.
+        confident_thresholds = [np.mean(pred_probs[:, k][k_in_l[k]]) for k in unique_classes]
+    else:
+        confident_thresholds = np.array(
+            [np.mean(pred_probs[:, k][labels == k]) for k in range(pred_probs.shape[1])]
```
Comment (jwmueller) [inline cleanlab/count.py:1135]: replace `range(pred_probs.shape[1])` with `unique_classes`?

File: cleanlab/count.py [hunk 9]
```diff
@@ -1108,12 +1107,31 @@ def get_confident_thresholds(labels: np.array, pred_probs: np.array) -> np.array
...
     confident_thresholds : np.array
       An array of shape ``(K,)``.
     """
-    confident_thresholds = np.array(
-        [np.mean(pred_probs[:, k][labels == k]) for k in range(pred_probs.shape[1])]
-    )
+
+    if multi_label:
+        # [0, 1, 2, ... num_classes - 1] Assumes all classses are represented in labels
+        unique_classes = range(pred_probs.shape[1])
```
Comment (jwmueller) [inline cleanlab/count.py:1128]: move this before the if statement?

File: cleanlab/count.py (hunk 2 above)
Comment (jwmueller) [reply cleanlab/count.py:1137]: Isn't the return type of this a list for the multilabel setting?  

Current docstring for the function says it returns np.array always

File: cleanlab/count.py [hunk 10]
```diff
@@ -1098,6 +1095,8 @@ def get_confident_thresholds(labels: np.array, pred_probs: np.array) -> np.array
     labels : np.array
       An array of shape ``(N,)`` of noisy labels, i.e. some labels may be erroneous.
       Elements must be in the set 0, 1, ..., K-1, where K is the number of classes.
+      All unique labels should be present, such that:
```
Comment (jwmueller) [inline cleanlab/count.py:1098]: wouldn't "all unique labels": be present by definition?
Maybe reword to: 

"all possible classes in 0, 1,...K-1" must be present in labels"

**Review by jwmueller — APPROVED** [review]
LGTM, just had minor comments.

We should consider also adding the clarification that "All possible classes in 0,1,...,K-1 must be present in labels" to the docstrings of higher-level functions like find_label_issues(). 

Can do this in separate follow-up PR though.

File: cleanlab/benchmarking/noise_generation.py [hunk 11]
```diff
@@ -116,7 +116,7 @@ def generate_noisy_labels(true_labels, noise_matrix):
     Parameters
     ----------
     true_labels : np.array
-      An array of shape ``(N,)`` representing perfect labels, without any
+      An array of shape ``(N, )`` representing perfect labels, without any
```
[the remark this answers is not in the document]
Comment (cgnorthcutt) [reply cleanlab/benchmarking/noise_generation.py:119]: this was convention at other places i've worked because it reduces edit distance if you need to add a second dimension. It's also clearer to have separation between two punctuation with different meaning.

[the remark this answers is not in the document]
Comment (cgnorthcutt) [reply cleanlab/benchmarking/noise_generation.py]: will change back to numpy convention to move things forward in agreement
